File: fabric-mod/src/main/java/com/mcgym/policy/Actions.java
package com.mcgym.policy;

// the trainers multi discrete action space, 22 logits split into 7 heads by BINS,
// argmax per head gives [forward, strafe, jump, sprint, yaw, pitch, attack]
public final class Actions {

    public static final int[] BINS = {3, 3, 2, 2, 5, 5, 2};

    private static final float[] FORWARD = {-1.0f, 0.0f, 1.0f};
    private static final float[] STRAFE  = {-1.0f, 0.0f, 1.0f};
    private static final float[] CAMERA  = {-10.0f, -3.0f, 0.0f, 3.0f, 10.0f};

    public float   forward;
    public float   strafe;
    public boolean jump;
    public boolean sprint;
    public float   yaw;
    public float   pitch;
    public boolean attack;

    public static Actions decode(float[] logits) {
        int[] idx = new int[BINS.length];
        int off = 0;
        for (int head = 0; head < BINS.length; head++) {
            idx[head] = argmax(logits, off, BINS[head]);
            off += BINS[head];
        }

        Actions a = new Actions();
        a.forward = FORWARD[idx[0]];
        a.strafe  = STRAFE[idx[1]];
        a.jump    = idx[2] != 0;
        a.sprint  = idx[3] != 0;
        a.yaw     = CAMERA[idx[4]];
        a.pitch   = CAMERA[idx[5]];
        a.attack  = idx[6] != 0;

        return a;
    }

    private static int argmax(float[] logits, int off, int len) {
        int best = 0;
        float val = logits[off];
        for (int i = 1; i < len; i++) {
            if (logits[off + i] > val) {
                val = logits[off + i];
                best = i;
            }
        }
        
        return best;
    }
}

File: trainer/mcgym/models/actions.py
"""multi discrete action space for the wood policy."""
from __future__ import annotations

import numpy as np

from mcgym.schema import spec

# heads forward strafe jump sprint yaw pitch attack, rest of ACTION_DTYPE stays zero
BINS = [3, 3, 2, 2, 5, 5, 2]

# per head value tables, index recovers the gym action value
FORWARD = np.array([-1.0, 0.0, 1.0], dtype=np.float32)
STRAFE = np.array([-1.0, 0.0, 1.0], dtype=np.float32)
JUMP = np.array([0, 1], dtype=np.uint8)
SPRINT = np.array([0, 1], dtype=np.uint8)
YAW = np.array([-10.0, -3.0, 0.0, 3.0, 10.0], dtype=np.float32)
PITCH = np.array([-10.0, -3.0, 0.0, 3.0, 10.0], dtype=np.float32)
ATTACK = np.array([0, 1], dtype=np.uint8)


def actions_to_records(idx: np.ndarray) -> np.ndarray:
    """map (N, 7) head indices to an (N,) ACTION_DTYPE array."""
    idx = np.asarray(idx)
    if idx.ndim != 2 or idx.shape[1] != len(BINS):
        raise ValueError(f"expected (N, {len(BINS)}) action indices, got {idx.shape}")
    out = np.zeros(idx.shape[0], dtype=spec.ACTION_DTYPE)
    out["forward"] = FORWARD[idx[:, 0]]
    out["strafe"] = STRAFE[idx[:, 1]]
    out["jump"] = JUMP[idx[:, 2]]
    out["sprint"] = SPRINT[idx[:, 3]]
    out["yaw_delta"] = YAW[idx[:, 4]]
    out["pitch_delta"] = PITCH[idx[:, 5]]
    out["attack"] = ATTACK[idx[:, 6]]
    return out
